presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
                     true));
         }
 
+        private Optional<DecorrelationResult> rewriteLimitWithRowCountGreaterThanOne(DecorrelationResult childDecorrelationResult, long limit)
+        {
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+
+            // rewrite Limit to RowNumberNode partitioned by constant symbols
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    idAllocator.getNextId(),
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.getConstantSymbols()),
+                    symbolAllocator.newSymbol("row_number", BigintType.BIGINT),
+                    Optional.of(toIntExact(limit)),
```

Comment:
maybe we should change `maxRowCountPerPartition` be an `Optional<Long>`?

Reply:
I think that the choice of Integer was intentional. It reflects the fact that:
`// Operators can handle MAX_VALUE rows per page ...` - as a comment in the class `WindowFilterPushDown` says.
It is also consistent with `TopNRowNumberNode` having `int maxRowCountPerPartition`.